Wrappers/Python/test/test_approximate_gradient.py

```diff
@@ -0,0 +1,284 @@
+            
+            
+
+    def test_approximate_gradient_not_equal_full(self):
+        self.assertFalse((self.f_stochastic.full_gradient(
+            self.initial) == self.f_stochastic.gradient(self.initial).array).all())
+
+    
+    def test_sampler(self):
+        self.assertTrue(isinstance(self.f_stochastic.sampler, SamplerRandom))
+        f = SGFunction(self.f_subsets)
+        self.assertTrue(isinstance(f.sampler, SamplerRandom))
+        self.assertEqual(f.sampler._type, 'random_with_replacement')
+
+    
+    def test_call(self):
+        self.assertAlmostEqual(self.f_stochastic(
+            self.initial), self.f(self.initial), 1)
+
+    
+    def test_full_gradient(self):
+        self.assertNumpyArrayAlmostEqual(self.f_stochastic.full_gradient(
+            self.initial).array, self.f.gradient(self.initial).array, 2)
+
+    
+    def test_value_error_with_only_one_function(self):
+        with self.assertRaises(ValueError):
+            SGFunction([self.f], self.sampler)
+            pass
+
+    
+    def test_type_error_if_functions_not_a_list(self):
+        with self.assertRaises(TypeError):
+            SGFunction(self.f, self.sampler)
+
+    
+    def test_sampler_without_next(self):
+        class bad_Sampler():
+            def init(self):
+                pass
+        bad_sampler = bad_Sampler()
+        with self.assertRaises(ValueError):
+            SGFunction([self.f, self.f], bad_sampler)
+
+    
+    def test_sampler_out_of_range(self):
+        bad_sampler = Sampler.sequential(10)
+        f = SGFunction([self.f, self.f], bad_sampler)
+        with self.assertRaises(IndexError):
+            f.gradient(self.initial)
+            f.gradient(self.initial)
+            f.gradient(self.initial)
+            
+    def test_partition_weights(self):
+        f_stochastic=SGFunction(self.f_subsets, Sampler.sequential(self.n_subsets))
+        self.assertListEqual(f_stochastic._partition_weights, [1 / self.n_subsets] * self.n_subsets)
+        with self.assertRaises(ValueError):
+            f_stochastic.set_data_partition_weights( list(range(self.n_subsets)))
+        with self.assertRaises(ValueError):
+            f_stochastic.set_data_partition_weights( [1])
+        with self.assertRaises(ValueError):
+            f_stochastic.set_data_partition_weights( [-1]+[2/(self.n_subsets-1)]*(self.n_subsets-1))
+        a=[i/float(sum(range(self.n_subsets))) for i in range(self.n_subsets)]
+        f_stochastic.set_data_partition_weights( a)
+        self.assertListEqual(f_stochastic._partition_weights, a )
+        f_stochastic.gradient(self.initial)
+        for i in range(1,20):
+            f_stochastic.gradient(self.initial)
+            self.assertEqual(f_stochastic.data_passes[i], f_stochastic.data_passes[i-1]+a[i%self.n_subsets])
+        
+        
+    
+    @unittest.skipUnless(has_astra, "Requires ASTRA GPU")
+    def test_SGD_simulated_parallel_beam_data(self):
```

Comment:
if it requires astra make sure the unit test name has `astra` in it. See https://github.com/TomographicImaging/CIL/pull/1729/files